github/commitstatus.go

```diff
@@ -262,35 +259,46 @@ func min(a, b int) int {
 //  2. The HTTP status code is in a retryable subset of the 5xx status codes.
 //     In this case, it returns the same as the input parameter waitTime.
 func checkForRetry(res httpResponse, waitTime, maxSleepTime, jitter time.Duration,
-) (time.Duration, string, error) {
+) (time.Duration, string) {
 	retryableStatusCodes := []int{
 		http.StatusInternalServerError, // 500
 		http.StatusBadGateway,          // 502
 		http.StatusServiceUnavailable,  // 503
 		http.StatusGatewayTimeout,      // 504
 	}
 
-	switch {
+	// Are we rate limited ?
 	// If the request exceeds the rate limit, the response will have status 403 Forbidden
 	// and the x-ratelimit-remaining header will be 0
 	// https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit
-	case res.statusCode == http.StatusForbidden && res.rateLimitRemaining == 0:
+	if res.statusCode == http.StatusForbidden && res.rateLimitRemaining == 0 {
 		// Calculate the sleep time based solely on the server clock. This is unaffected
 		// by the inevitable clock drift between server and client.
 		sleepTime := res.rateLimitReset.Sub(res.date)
+		// Be robust to possible races in the GitHub backend. This avoids failing too early.
+		if sleepTime < 0 {
+			sleepTime = 0
+		}
 		// Be a good netizen by adding some jitter to the time we sleep.
 		sleepTime += jitter
```

Comment:
I will check again tomorrow but I think your PR still doesn't fix the situation that I think it happend (which is actually the simplest possible case)

- first request failed. Sleep time was around 4min30sec
- so we sleep this time
- second request is done. lets assume `res.rateLimitReset.Sub(res.date)` returned 0.
- by pure luck jitter rand func also returned 0
- even if we had your current code, third attempt would never happen because we brake out of the loop

Reply:
You might not be talking to the same server on both attempts. They might have clocks off by just the right amount such that the second requests appears to arrive milliseconds before the reset.

But your point is still valid so probably making sure jitter is at least 1 would be an easy fix.

Reply:
@aliculPix4D 
I don't understand this scenario. Now that we have the dedicated unit test, can you come up with a test that shows the bug? If not, I think it is better to have a face chat so that I can understand the details.

Reply:
> so probably making sure jitter is at least 1 would be an easy fix.

If the scenario turns out to be possible, I agree that this for sure is an easy fix.

Reply:
I would suggest you also change the jitter time to:
`5*time.Second + time.Duration(rand.Intn(30)) * time.Second`
https://github.com/Pix4D/cogito/blob/389e9a79e6622c6edc9fcdeb6119122a33b4db97/cogito/ghcommitsink.go#L43

Reply:
As discussed face-to-face, I did not change the jitter on purpose. The code must be correct also if jitter equals 0.